Implement SafeAreaEdges property and per-edge safe area control for Android

NirmalKumarYuvaraj opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

<!--
!!!!!!! MAIN IS THE ONLY ACTIVE BRANCH. MAKE SURE THIS PR IS TARGETING MAIN. !!!!!!! 
-->

### Description of Change

<!-- Enter description of the fix in this section -->
This pull request introduces a unified and optimized approach for handling safe area insets and keyboard visibility across Android platform views in the codebase. The changes add a new `SafeAreaHandler` class, refactor `ContentViewGroup` and `LayoutViewGroup` to use this handler, and implement supporting types and extension methods for safe area calculations. The main goal is to ensure consistent, performant, and keyboard-aware safe area handling for layouts, reducing repeated hierarchy traversals and improving maintainability.

**Safe Area Handling Improvements**

* Added the new `SafeAreaHandler` class (`SafeAreaHandler.cs`) to encapsulate shared logic for safe area and keyboard inset handling, including caching, hierarchy traversal optimizations, and a unified window insets listener.
* Refactored `ContentViewGroup` and `LayoutViewGroup` to instantiate and use `SafeAreaHandler`, including setup of window insets listeners and safe area adjustment logic in their constructors and `OnLayout` methods. [[1]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR17-R58) [[2]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R19-R65) [[3]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR116-R121) [[4]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R129-R134)

**Supporting Types and Extensions**

* Introduced the `SafeAreaPadding` record struct and `WindowInsetsExtensions` static class for safe area and keyboard inset calculations, conversion, and rectangle adjustment.

**Dependency Updates**

* Added `AndroidX.Core.View` using statements to affected files to support new window insets APIs. [[1]](diffhunk://#diff-5fdf872b32d4f8ee5971be75e1e5f45a53ec6c55fbdc701b54e02a175ca05ccaR7) [[2]](diffhunk://#diff-3f4ba0ef39c71f2f98bc747387389cd17c63805a37849606234f7d8eb5d42e17R7)

These changes collectively provide a robust foundation for safe area and keyboard-aware layout on Android, making it easier to maintain and extend in the future.

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes # https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2433909

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->
### Output

https://github.com/user-attachments/assets/4785f4c6-49ca-4f4e-9feb-519277b26942

NirmalKumarYuvaraj committed:
Implemented safearea handling for android

NirmalKumarYuvaraj committed:
minor changes

NirmalKumarYuvaraj committed:
code updated

albyrock87 commented:
Thank you so much for this effort!
I have some concerns regarding performance given this will be called on every node and it involves a lot of JNI calls.
Do you think we could move the majority of the logic on the Java side?

albyrock87 commented on src/Core/src/Platform/Android/SafeAreaHandler.cs:
```diff
@@ -0,0 +1,384 @@
+                currentDepth++;
+
+                // Check if parent is a ContentViewGroup or LayoutViewGroup that responds to safe area
+                if (parent is ContentViewGroup parentContentGroup)
```
What about having an interface implemented by both `ContentViewGroup` and `LayoutViewGroup` so that we can have this `Quick check` code only once?

albyrock87 replied:
I think we already have one, https://github.com/dotnet/maui/blob/main/src/Core/src/Core/ICrossPlatformLayoutBacking.cs

NirmalKumarYuvaraj replied:
@albyrock87 , I have updated the code changes. If you have any concerns please let me know

NirmalKumarYuvaraj committed:
code changes

NirmalKumarYuvaraj committed:
optimized safearea

NirmalKumarYuvaraj committed:
scrollview supports safearea edges

NirmalKumarYuvaraj committed:
Updated code changes

NirmalKumarYuvaraj committed:
fixed contentpage issue

NirmalKumarYuvaraj committed:
keyboard fix

NirmalKumarYuvaraj committed:
Used consumption logic

jsuarezruiz commented:
/azp run

jsuarezruiz reviewed (changes requested):
These are the tests to validate the changes, right?
https://github.com/dotnet/maui/pull/30908/

Could we expand it to support more edge cases? I would include more tests like:

- `SafeAreaBorderOrientationPortraitToLandscape` - Tests orientation change from portrait to landscape, verifying that the safe area adapts, the border dimensions change, and insets adjust on notched devices.
- `SafeAreaBorderOrientationLandscapeToPortrait` - Tests orientation change from landscape to portrait, ensuring the safe area reverts correctly and border dimensions differ between orientations.
- `SafeAreaBorderMultipleOrientationChanges` - Tests multiple sequential orientation changes (portrait to landscape and back), verifying border visibility, positioning consistency within the same orientation (with tolerance), and overall stability.
- `SafeAreaBorderSoftInputBehavior` - Tests soft input (keyboard) behavior with bottom edge set to SoftInput, checking that the border shrinks when the keyboard appears and returns to original size after dismissal.
- `SafeAreaBorderSoftInputWithOrientationChange` - Tests soft input with orientation change (portrait to landscape) while keyboard is visible, ensuring the border maintains visibility and dimensions differ between orientations.

NirmalKumarYuvaraj committed:
optimized approach

NirmalKumarYuvaraj committed:
more changes

NirmalKumarYuvaraj commented on src/Core/src/Platform/Android/SafeAreaHandler.cs:
```diff
@@ -0,0 +1,248 @@
+        new WindowInsetsListener(this, () => _owner.RequestLayout());
+    void InvalidateSafeArea()
+    {
+        _safeAreaInvalidated = true;
```
@jsuarezruiz , Modified the code changes

NirmalKumarYuvaraj commented on src/Core/src/Platform/Android/SafeAreaPadding.cs:
```diff
@@ -0,0 +1,90 @@
+	}
+
+	public AInsets ToInsets() =>
+		AInsets.Of((int)Left, (int)Top, (int)Right, (int)Bottom) ?? AInsets.Of(0, 0, 0, 0)!;
```
@jsuarezruiz , Removed the code. since it was not used anywhere

PureWeen commented:
/azp run

PureWeen reviewed (changes requested):
I still need to review this a little bit deeper but I do still wonder about @albyrock87 's comment here

https://github.com/dotnet/maui/pull/31022#discussion_r2268725703

I wonder about the approach here with storing the safea area into separate handlers and then modifying the arrange pass. 

In theory could we just have a global handler that sets the padding on the view? 

Basically what I've done to fix the safearea on various views for edge to edge is just set the padding inside the windowinset listener

<img width="803" height="499" alt="image" src="https://github.com/user-attachments/assets/b70763bf-3391-410c-90f8-538ff7e39561" />

could we take the same approach here? Like, we just have a window listener applied to ContentViewGroup and LayoutViewGroup and then inside the "ApplyWindowLIstener" it extracts the safeareas and sets the padding on the views.

If the padding is set on the views than the measure and arrange passes will just work. The main reason we don't do this approach on iOS (AFAIK) is that iOS doesn't really have a concept of "padding" you can set on views, you always just fake the padding inside LayoutSubviews.

NirmalKumarYuvaraj committed:
Added more test cases

NirmalKumarYuvaraj committed:
more changes

NirmalKumarYuvaraj committed:
Test failure fixes

NirmalKumarYuvaraj committed:
more refactor

NirmalKumarYuvaraj committed:
implemented Global listener

# Conflicts:
#	src/Core/src/Platform/Android/LayoutViewGroup.cs

NirmalKumarYuvaraj committed:
Resolved issues

NirmalKumarYuvaraj committed:
updated review corrections

NirmalKumarYuvaraj commented on src/Core/src/Platform/Android/MauiAppCompatActivity.cs:
```diff
@@ -9,6 +9,16 @@ namespace Microsoft.Maui
+		/// Gets the shared GlobalWindowInsetListener instance for this activity.
+		/// This ensures all views use the same listener instance for coordinated inset management.
+		/// </summary>
+		public GlobalWindowInsetListener GlobalWindowInsetListener =>
```
@PureWeen , Done, marked it as internal.

NirmalKumarYuvaraj committed:
modal listener

NirmalKumarYuvaraj committed:
ScrollviewHandling

NirmalKumarYuvaraj committed:
more changes

jsuarezruiz reviewed (changes requested):
There are several testa failing, mostly all the Android tests using snapshots. The PR fundamentally changes how MAUI calculates and applies padding/margins for system UI elements. This is a breaking visual change that requires changes.
Currently we are cutting 60 top px on Android: https://github.com/dotnet/maui/blob/6d868559b690edcd60542b3c174c82f150fdab90/src/Controls/tests/TestCases.Shared.Tests/UITest.cs#L296


Example of a failing test, FontFamily:
`Snapshot different than baseline: LabelUITests_FontFamily_FontAwesome.png (17.88% difference)`
<img width="1080" height="1735" alt="image" src="https://github.com/user-attachments/assets/5711a8b6-2efb-4cb3-99fd-a2e192d188e7" />

Next, the differences: 
<img width="1080" height="1735" alt="image" src="https://github.com/user-attachments/assets/948f0952-e49d-4af7-99de-042d8756a342" />

NirmalKumarYuvaraj committed:
Implemented GetDisplayDensity

NirmalKumarYuvaraj committed:
Addressed test failures

NirmalKumarYuvaraj committed:
Addressed test case failures

jsuarezruiz commented:
/azp run

PureWeen commented on src/Controls/src/Core/SafeAreaElement.cs:
```diff
@@ -19,6 +22,22 @@ static void OnSafeAreaEdgesChanged(BindableObject bindable, object oldValue, obj
+
+#if ANDROID
+				// On Android, request layout does not call OnApplyWindowInsets. so we manually call it.
+				if (bindable is Element element && element.Handler is IElementHandler handler)
```
I think you could achieve this with a mapper couldn't you? 

Add a mapper to MapSafeAreaEdges and have a specific one for android that calls this code?

The InvalidateMeasure should probably also be in a mapper

PureWeen committed:
Refactor GlobalWindowInsetListener handling for MauiScrollView to avoid redundant listener setup

NirmalKumarYuvaraj committed:
scroll view changes

NirmalKumarYuvaraj committed:
added intersecting logic for scrollview

jsuarezruiz commented:
/azp run

NirmalKumarYuvaraj committed:
more changes

NirmalKumarYuvaraj committed:
more changes

NirmalKumarYuvaraj committed:
intersect logic working

NirmalKumarYuvaraj committed:
final commit

PureWeen commented:
/rebase

PureWeen commented:
/azp run

PureWeen commented:
/rebase

PureWeen committed:
- fix intersection calculation code

PureWeen committed:
turn off decor for all APIs

PureWeen committed:
- set to container

jsuarezruiz reviewed (dismissed):
Build FAILED in .NET Standard:

```
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,25): error CS0117: 'OperatingSystem' does not contain a definition for 'IsIOS' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.1]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,53): error CS0117: 'OperatingSystem' does not contain a definition for 'IsMacCatalyst' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.1]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,25): error CS0117: 'OperatingSystem' does not contain a definition for 'IsIOS' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.0]
/Users/builder/azdo/_work/1/s/src/Controls/src/Core/ContentPage/ContentPage.cs(183,53): error CS0117: 'OperatingSystem' does not contain a definition for 'IsMacCatalyst' [/Users/builder/azdo/_work/1/s/src/Controls/src/Core/Controls.Core.csproj::TargetFramework=netstandard2.0]
    4 Error(s)

```

NirmalKumarYuvaraj committed:
APi 30 test failures addressed

PureWeen committed:
- fix shell

PureWeen committed:
- appbar layout

PureWeen committed:
- remove listener

PureWeen committed:
- fix shell flyout

PureWeen committed:
- remove from the root view

PureWeen committed:
- additional updates

PureWeen committed:
- fix on root

PureWeen committed:
- fix up tests

PureWeen committed:
- fix up android tests

PureWeen committed:
- account for bottom tabs

PureWeen committed:
- fix orientation changes

# Conflicts:
#	src/Controls/tests/TestCases.Android.Tests/snapshots/android/BorderBackgroundSizeUpdatesWhenRotatingScreen.png
#	src/Controls/tests/TestCases.Android.Tests/snapshots/android/ButtonsLayoutResolveWhenParentSizeChangesSizeButtonsDownLandscape.png
#	src/Controls/tests/TestCases.Android.Tests/snapshots/android/CarouselViewItemShouldScaleProperly.png

PureWeen commented:
/azp run MAUI-UITests-public

PureWeen committed:
- fix insets on shell flyout

PureWeen committed:
- fix up android safearea insets

PureWeen committed:
- more fixes

PureWeen commented:
/azp run

PureWeen commented on src/TestUtils/src/UITest.Appium/Actions/AppiumScrollActions.cs:
```diff
@@ -287,7 +287,7 @@ void ScrollToUp(AppiumDriver driver, AppiumElement element, ScrollStrategy strat
 			AppiumElement? result = ScrollUpUntilPresent(driver, target, element, strategy, swipePercentage, swipeSpeed, withInertia);
 
-			while (result is not null)
+			while (result is null)
```
:-/

PureWeen committed:
- fix resetting of inserts if there's no intersection

PureWeen commented:
/azp run

PureWeen committed:
- fix to use measured height and width instead

PureWeen commented:
/azp run

PureWeen committed:
- fix runner to rerun tests more correctly

PureWeen committed:
- fix

PureWeen commented:
/azp run

PureWeen committed:
- fix device tests

PureWeen committed:
- update

PureWeen commented:
/azp run

PureWeen committed:
- fix flyout paddings

PureWeen committed:
- faith is a facet

PureWeen commented:
/azp run

PureWeen committed:
- adjust device tests to account for padding

PureWeen commented:
/azp run

PureWeen committed:
- fix

PureWeen committed:
- fixes

PureWeen commented:
/azp run

PureWeen committed:
Increase epsilon value in ShellFlyoutTests

PureWeen commented:
/azp run

PureWeen committed:
Change epsilon value in ShellFlyoutTests

Updated epsilon value for header position assertion in tests.

PureWeen committed:
Update ShellFlyoutTests.cs

PureWeen commented:
/backport to release/10.0.1xx-rc2

PureWeen committed:
Refactor platform-specific safe area insets handling

PureWeen committed:
Add AndroidX.Core.View import for ShellFlyoutTests

PureWeen commented:
/backport to release/10.0.1xx-rc2

PureWeen committed:
Fix formatting in ShellFlyoutTests.cs

PureWeen commented:
/azp run

PureWeen commented:
/backport to release/10.0.1xx-rc2

PureWeen committed:
Improve assertion message and add conditional return

Updated assertion message for clarity and added conditional compilation for return statement.

PureWeen committed:
Document test behavior for Android headless mode

Added comments regarding test behavior differences between headless and non-headless modes for Android.

PureWeen commented:
/backport to release/10.0.1xx-rc2

PureWeen committed:
Clean up ShellFlyoutTests by removing comments

Removed commented-out code regarding test behavior for headless vs non-headless view positioning. Updated test method for better clarity.

PureWeen committed:
Refactor ShellFlyoutTests for improved readability

PureWeen committed:
Update ShellFlyoutTests.cs

PureWeen committed:
Add Theory attribute to FlyoutHeaderScroll test

PureWeen commented:
/azp run
